### PR TITLE
fix(22547): fix e2e test import token and prevent interaction before modal is closed

### PR DESCRIPTION
**Pull request by DDDDDanica**
## **Description**
There are two issues we are resolving here:

1. For failed **`allows importing multiple tokens from search`**
- When we are on mainnet, `isTokenDetectionInactiveOnMainnet` is `false` for selector `getTokenList`
- Hence we will be looking for `state.metamask.tokenList`, while that results in {} in our tests builder
- We performs the switch network action, while will result in `isTokenDetectionInactiveOnMainnet `as `true` and we will use the hardcoded `STATIC_MAINNET_TOKEN_LIST`
- And current component is missing `componentWillReceiveProps` lifecycle methods to handle props change
- Therefore we didn’t pass the new props to `token-search.component`
- Hence the solution from my side is to refactor the component into a functional component and track the props change by `useEffect`

2. For **preventing interactions between network switching**, starting on `Ganache/Localhost`
- Click the network picker
- The "Select a network" modal appears
- Click "Ethereum Mainnet"
- The modal closes, but we're still on Ganache/Localhost and everything is still clickable, including the "Import tokens" button
- A few moments later, the UI updates to show the new network, and a "Connecting to _____" overlay appears, which prevents interactions until the network has loaded

For the above scenario, we should prevent use from interacting before network is successfully switched.

## **Related issues**

Fixes: https://app.zenhub.com/workspaces/extension-delivery-board-6216892781ac020010e826d1/issues/gh/metamask/metamask-extension/22547

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Review by vthomas13 — DISMISSED** [review]

File: test/e2e/tests/import-tokens.spec.js
Comment (DDDDDanica) [inline test/e2e/tests/import-tokens.spec.js:36]: Wait until 25eth is loaded from main network

File: test/e2e/tests/import-tokens.spec.js
Comment (DDDDDanica) [inline test/e2e/tests/import-tokens.spec.js:32]: temporarily fix to wait until network is switched

File: test/e2e/webdriver/driver.js
Comment (bergeron) [inline test/e2e/webdriver/driver.js:643]: If this ends up being the fix, it can be configured on the specific test instead:

```
await withFixtures({
  ...,
  failOnConsoleError: false,
}
```

**Comment by HowardBraham** [comment]
@DDDDDanica hmm this is unfortunately failing test-e2e-chrome-multichain

File: test/e2e/tests/import-tokens.spec.js
Comment (Gudahtt) [inline test/e2e/tests/import-tokens.spec.js:32]: What is the purpose of finding the currency?

File: test/e2e/webdriver/driver.js
Comment (Gudahtt) [reply test/e2e/webdriver/driver.js:643]: Yeah, we definitely don't want to remove this line. This setting is supposed to make tests fail when we see an error message.

File: ui/components/multichain/network-list-menu/network-list-menu.js
Comment (Gudahtt) [inline ui/components/multichain/network-list-menu/network-list-menu.js:344]: Hmm, it's unclear how this relates to the problem. I know you mentioned in Slack that this had changed recently, but the previous version behaved the same way (the toggle happened first, see here: https://github.com/MetaMask/metamask-extension/blame/e1da0ef183279fcb5c4e2b64f7696533d3486ac0/ui/components/multichain/network-list-menu/network-list-menu.js#L344)

In both cases, the modal is closed before the switch completes, because we're not `await`iing the call to switch networks.

File: ui/components/app/import-token/token-search/token-search.component.js
Comment (Gudahtt) [inline ui/components/app/import-token/token-search/token-search.component.js:1]: Perhaps this component conversion could be done in a separate PR from the test changes? They seem independent of each other.

File: test/e2e/webdriver/driver.js
Comment (DDDDDanica) [reply test/e2e/webdriver/driver.js]: yes, i figured out a mock now and can bring it back, see [567583](https://github.com/MetaMask/metamask-extension/pull/22548/commits/567583ece63b2f4538040c54e669a308d21bbf63)

File: test/e2e/tests/import-tokens.spec.js
Comment (DDDDDanica) [reply test/e2e/tests/import-tokens.spec.js:62]: This is to wait until network is switched. when it is still switching, we will see a loader and no secondary value. here's the temp solution to make sure we wait until network is switched then we click "import" button

**Comment by DDDDDanica** [comment]
> @DDDDDanica hmm this is unfortunately failing test-e2e-chrome-multichain

yeah there's some UI difference for multi-chain, i added this helper to tell difference https://github.com/MetaMask/metamask-extension/blob/567583ece63b2f4538040c54e669a308d21bbf63/test/e2e/tests/import-tokens.spec.js#L60

File: ui/components/app/import-token/token-search/token-search.component.js
Comment (DDDDDanica) [reply ui/components/app/import-token/token-search/token-search.component.js:1]: IMO It should be within the fix, it will handle the props change and load the list for component again.

File: test/e2e/tests/import-tokens.spec.js
Comment (DDDDDanica) [reply test/e2e/tests/import-tokens.spec.js:53]: It is a good idea, i can see many places in e2e test using this class. Created an issue in delivery board to track:
https://app.zenhub.com/workspaces/extension-delivery-board-6216892781ac020010e826d1/issues/gh/metamask/metamask-extension/22552

File: ui/components/multichain/network-list-menu/network-list-menu.js
Comment (DDDDDanica) [reply ui/components/multichain/network-list-menu/network-list-menu.js]: reverted [here](05bf3655077d7d9641663f51af1258b56b2bfab4)

File: test/e2e/tests/import-tokens.spec.js
Comment (bergeron) [inline test/e2e/tests/import-tokens.spec.js:18]: Mocking seems reasonable, since I see other e2e tests mocking some network calls.  Although this could break if the token list changes such that the first 3 tokens starting with `cha` become different.

If we decide these should be real calls, we could either:
1. Select the 1st and 3rd tokens in the list, since it's currently the 2nd one that has no prices.  (This is also fragile on the token list changing)
2. Pass `failOnConsoleError: false` to `withFixtures`

Altough a token not having a price should probably be a `console.warn` instead of `console.error` anyway.

Comment (HowardBraham) [reply test/e2e/tests/import-tokens.spec.js:18]: I agree that not having a price should probably be a `console.warn` instead of `console.error`

I also think that in the longer-term, we should fix all our console errors and convert all tests to `failOnConsoleError: true`

Comment (DDDDDanica) [reply test/e2e/tests/import-tokens.spec.js:18]: There was a fix recently about this `console.error` in assets-controller: https://github.com/MetaMask/core/blob/main/packages/assets-controllers/src/token-prices-service/codefi-v2.ts#L378.

After thinking again about this test, what i have in mind for this e2e test is to return `{}` for any query, which is the response from current API (GET `https://price-api.metafi.codefi.network/v2/chains/1/spot-prices?tokenAddresses=0x06af07097c9eeb7fd685c692751d5c66db49c215%2C0x514910771af9ca656af840dff83e8264ecf986ca%2C0x7d4b8cce0591c9044a22ee543533b72e976e36c3&vsCurrency=ETH`). And UI should be able to handle this as well if the result is empty. what do you guys think?

File: test/e2e/tests/import-tokens.spec.js
Comment (DDDDDanica) [reply test/e2e/tests/import-tokens.spec.js]: Updates: i changed the way and used `network-display` to assert so we are not differentiating MULTICHAIN with this one. [cdb6626](https://github.com/MetaMask/metamask-extension/pull/22548/commits/cdb6626a8828fee737cc23f3face0904ddc519ab)

**Comment by DDDDDanica** [comment]
@metamaskbot update-policies

File: test/e2e/tests/import-tokens.spec.js
Comment (Gudahtt) [reply test/e2e/tests/import-tokens.spec.js]: Hmm, I still don't follow why we need any wait step here. When the loader is on the screen, the "Import tokens" button is not clickable, so the step where that button is clicked should suffice on its own. The `clickElement` function waits until the target is both located and clickable.

File: test/e2e/tests/import-tokens.spec.js
Comment (Gudahtt) [reply test/e2e/tests/import-tokens.spec.js:19]: There are cases where it's useful to print an error to the console in production. We could use warn for those cases, but it's nice to be able to distinguish between operational errors and warnings.

Maybe instead of getting rid of `failOnConsoleError` completely, we could add a configuration for test-specific expected errors.

Comment (Gudahtt) [reply test/e2e/tests/import-tokens.spec.js:19]: Returning an empty object makes sense! Not sure why the API is returning that at the moment, that doesn't seem right. But our code handles it fine seemingly.

File: test/e2e/tests/import-tokens.spec.js
Comment (Gudahtt) [inline test/e2e/tests/import-tokens.spec.js:66]: The text for this label might change in between these two lines. To verify that an element has certain text content, you should build the expected text into the locator to prevent a race condition.

File: ui/components/app/import-token/token-search/token-search.component.js
Comment (Gudahtt) [reply ui/components/app/import-token/token-search/token-search.component.js:1]: Right, I don't mean that it's unrelated, just that it's independent. In general it is a good practice to make independent changes in separate PRs.

File: test/e2e/tests/import-tokens.spec.js
Comment (Gudahtt) [reply test/e2e/tests/import-tokens.spec.js]: Oh right, it's because the loading overlay _isn't_ up initially, and that's still the case in this PR. Waiting for confirmation of network switch makes sense then.

File: test/e2e/tests/import-tokens.spec.js
Comment (Gudahtt) [reply test/e2e/tests/import-tokens.spec.js:66]: Making this change should let us get rid of that 1 second delay as well

File: test/e2e/tests/import-tokens.spec.js
Comment (DDDDDanica) [reply test/e2e/tests/import-tokens.spec.js:19]: It is included in [96f1d6](https://github.com/MetaMask/metamask-extension/pull/22548/commits/96f1d66be8d32b1a60aafb27bd64d722d876ad06)

File: ui/components/app/import-token/token-search/token-search.component.js
Comment (Gudahtt) [inline ui/components/app/import-token/token-search/token-search.component.js:41]: Hmm, why is `searchQuery` in this list of dependencies? I don't think we need to reconstruct this when the query changes

File: ui/components/app/import-token/token-search/token-search.component.js
Comment (Gudahtt) [inline ui/components/app/import-token/token-search/token-search.component.js:69]: Using a wrapper anonymous function here kinda defeats the purpose of using a `useCallback` hook for this handler. `useCallback` prevents the function reference from changing each render, but that won't matter if we're wrapping it in an anonymous function that itself changes references each render.

Maybe better to drop the `useCallback` altogether, and leave this behaving as it did before. But if you want to avoid some additional renders, you'll want to drop this wrapper anonymous function, and the one used for `clear` (and introduce a `useCallback` call for `clear`).

File: ui/components/app/import-token/token-search/token-search.component.js
Comment (Gudahtt) [inline ui/components/app/import-token/token-search/token-search.component.js:50]: Perhaps we can use `newSearchQuery` here, which would let us drop `searchQuery` as a dependency.

We know in this case that `newSearchQuery` is always equivalent to `searchQuery`. Dropping that dependency would let us preserve this callback when the query changes.

File: ui/components/app/import-token/token-search/token-search.component.js
Comment (Gudahtt) [reply ui/components/app/import-token/token-search/token-search.component.js:69]: Dropping `useCallback` would be OK because this would be a very minor optimization, probably not impactful. And because this will rerender very often anyway no matter what we do, because `tokenList` is from background state, and the way state syncing works today results in every state update updating all object references, causing rerenders of anything referencing background state.

Comment (DDDDDanica) [reply ui/components/app/import-token/token-search/token-search.component.js:66]: I added the drop [here](https://github.com/MetaMask/metamask-extension/pull/22548/commits/83b4a7063250507068f252348bb54534117eeab3) and will do another MR to do improvement of this component

File: ui/components/app/import-token/token-search/token-search.component.js
Comment (DDDDDanica) [reply ui/components/app/import-token/token-search/token-search.component.js]: Good idea, will address it in the next PR, I'm thinking we should merge this one to unblock other pipelines and improve what was left here in another one, wdyt?

**Review by shane-t — DISMISSED** [review]

File: ui/components/app/import-token/token-search/token-search.component.js
Comment (Gudahtt) [reply ui/components/app/import-token/token-search/token-search.component.js]: However this suggestion about `newSearchQuery` is no longer relevant now that `useCallback` isn't used anymore.

Comment (DDDDDanica) [reply ui/components/app/import-token/token-search/token-search.component.js]: Still a good idea to refactor this one. addressed [here](https://github.com/MetaMask/metamask-extension/pull/22548/commits/978bf0b881bd349eee7107c81cecc0068a636070)

File: ui/components/app/import-token/token-search/token-search.component.js
Comment (DDDDDanica) [reply ui/components/app/import-token/token-search/token-search.component.js:66]: changed [here](https://github.com/MetaMask/metamask-extension/pull/22548/commits/978bf0b881bd349eee7107c81cecc0068a636070)

File: ui/components/app/import-token/token-search/token-search.component.js
Comment (DDDDDanica) [reply ui/components/app/import-token/token-search/token-search.component.js]: Ah my bad, was debugging the component. removed [here](https://github.com/MetaMask/metamask-extension/pull/22548/commits/978bf0b881bd349eee7107c81cecc0068a636070)

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Review by brad-decker — APPROVED** [review]

File: test/e2e/tests/import-tokens.spec.js
Comment (DDDDDanica) [reply test/e2e/tests/import-tokens.spec.js:53]: Captured [here](https://github.com/MetaMask/metamask-extension/pull/22668) :)

**Pull request merged**
